util/src/main/java/tc/oc/pgm/util/usernames/BukkitUsernameResolver.java

```diff
@@ -0,0 +1,54 @@
+package tc.oc.pgm.util.usernames;
+
+import java.time.Instant;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import org.bukkit.Bukkit;
+import org.bukkit.OfflinePlayer;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+import tc.oc.pgm.util.bukkit.BukkitUtils;
+import tc.oc.pgm.util.nms.NMSHacks;
+
+public final class BukkitUsernameResolver extends AbstractUsernameResolver {
+  private static final SyncExecutor EXECUTOR = new SyncExecutor();
+  private static final Plugin PGM = BukkitUtils.getPlugin();
+
+  @Override
+  protected Executor getExecutor() {
+    return EXECUTOR;
+  }
+
+  @Override
+  protected void process(UUID uuid, CompletableFuture<UsernameResponse> future) {
+    future.complete(resolveUser(uuid));
+  }
+
+  private UsernameResponse resolveUser(UUID uuid) {
+    OfflinePlayer pl = Bukkit.getOfflinePlayer(uuid);
+    if (pl.getName() != null) {
+      long lastPlayed = pl.getLastPlayed();
+      return UsernameResponse.of(
+          pl.getName(),
+          lastPlayed > 0 ? Instant.ofEpochMilli(lastPlayed) : Instant.now(),
+          BukkitUsernameResolver.class);
+    }
+
+    // Does vanilla user cache hold this player?
+    String name = NMSHacks.getPlayerName(uuid);
+    if (name != null) {
+      return UsernameResponse.of(name, Instant.now(), BukkitUsernameResolver.class);
+    }
+    return UsernameResponse.empty();
+  }
+
+  private static class SyncExecutor implements Executor {
+
+    @Override
+    public void execute(@NotNull Runnable command) {
+      if (Bukkit.isPrimaryThread()) command.run();
+      else Bukkit.runOnMainThread(PGM, false, command);
```

Comment:
This won't work on non sportpaper
 https://github.com/Electroid/SportPaper/blob/main/patches/api/0080-Mojang-task-queue-access.patch#L27

Reply:
modified to use the NMSHacks method instead